Improve description of `locals` on guides/middleware (#5522)
withastro/docs: src/content/docs/en/guides/middleware.mdx

    ```js title="src/middleware.js"
    export function onRequest ({ locals, request }, next) {
        // intercept data from a request

>         // optionally, share data with the rendering process by modifying `locals`

Reviewer: ```suggestion
        // optionally, modify the data stored in `locals`
```
While yours may be the more technically accurate description, I don't think this exactly hits the mark on the helpfulness scale.

"Transform some data" is really clear to a reader that you're going to change the value of something. "Share data with a process" really doesn't convey with the same clarity the idea that middleware is used to change an object that is then available somewhere else.

Can we find a middle ground here? Is it "good enough" just to do the above?

Author: I agree that my wording was _too_ technical, but it is in line with the language just above it:

> This `locals` object is forwarded across the request handling process and is available as a property to `APIContext` and `AstroGlobal`.

I think a middle ground would be best. The original "transform the response" is what I thought was misleading since it may not affect the response at all, and there is another place to transform the actual `Response` object (after the `next` call), so it could be confusing.

I like your suggestion, we can go with it. I'd prefer "modify the _properties_" instead of "modify the _data_" just because you can set functions in there, but this is quite minor, WDYT?

Reviewer: Totally cool with "modify the properties" !

        locals.title = "New title";

        // return a Response or the result of calling `next()`

>         // optionally, transform the response by modifying the result

Reviewer: This doesn't feel well-placed to me, and in fact I don't even think it's necessary at all, if the we get the above code comment just right. The existing comment here is meant to highlight that you have to RETURN something, either `next()`  or a `Response` directly. I think this statement here only adds extra information gets (literally!) in the way of the point it's trying to make.

Is this also in fact correct? is the modifying done after what is returned in this function? Isn't returning *something* the final step here? So, if there are in fact two code comments, it seems to me like the one about what is returned by the middleware should go last.

I can think of some options off the top of my head, depending on what's most accurate, that I think feel more *helpful* to the reader:

1. change the existing statement to include the idea that the data, potentially modified, is passed along in the thing that's returned. e.g. "return a Response or the result of calling `next`, passing along `context.locals`"
2. swap the order of these, and reword the first one include information you believe is truly useful to include.
3. delete this, because it probably doesn't need to be spelled out HERE that the new info is available if the code comment above clearly explains the purpose behind transforming some data to mutate and pass along `context.locals`.

Author: The distinction is that:
- You can modify the `locals` to affect how the page/endpoint _will_ behave
- You can modify the return of `next` _after_ the page/endpoint has already completed

There are examples of this in the same page. For the first case, there is an example of changing the title inside `locals` allowing the page to render it differently. For the second, there is an example of redacting sensitive information by replacing segments of the resulting HTML.

Author: Taking another look at it, I agree that it gets in the way of the most essential information. Since there is already another section explaining about modifying the result of `next` I'll remove this comment.

Reviewer: That works for me!

### Storing data in `context.locals`

> `context.locals` is an object that can be manipulated inside the middleware. Adapters _might_ add data to it.

Reviewer: "Adapters might add data to it" is a little too vague to be helpful.

It's not clear whether the reader should take this as some kind of warning (i.e. if you're writing your own middleware , beware that there might be something else that's also manipulating the data in `context.locals`), or it's being used as an example of something that manipulates `context.locals` to provide real-usage context, or something else entirely.... fun fact!  Did you know...! :smile: 

Can you reword that statement in a way that makes it clear *why* they are being told that?

Author: I think this could be added on a tip. What do you think of removing that last sentence and adding this:

```md
:::tip[Adapter-provided properties]
Adapters may provide extra runtime information and functionality through the `locals` object. For example, SSR adapters may include information about the request or runtime.
:::
```

We can link to some examples. Of the official integrations, 4 modify the `locals` object.

Reviewer: I still don't know **why** you're telling me this, though? Again, it's not clear why I should care about this, whether it's something for me to watch out for...

Author: I think it is a mix of everything. It is something to be aware:
- An integration may provide something that you are checking manually
- Something you write might override a value used by an integration
- Information that would be hard to access or infer otherwise might already be included there (like CloudFlare's runtime)

Reviewer: OK, gotcha! No, I don't want examples here to clutter this up. The reasoning/why I should care is way more important. I think what you mentioned above is the kind of content that would be good to go inside a note!

Author: I sent a new version with a different tip. WDYT?

This `locals` object is forwarded across the request handling process and is available as a property to [`APIContext`](/en/reference/api-reference/#contextlocals) and [`AstroGlobal`](/en/reference/api-reference/#astrolocals) . This allows data to be shared between middlewares, API routes, and `.astro` pages. This is useful for storing request-specific data, such as user data, across the rendering step.

You can store any type of data inside `locals`: strings, numbers, and even complex data types such as functions and maps.

```js title="src/middleware.js"
export function onRequest ({ locals, request }, next) {
    // intercept data from a request

>     // optionally, share data with the rendering process by modifying `locals`
>     locals.user.name = "John Wick";
>     locals.welcomeTitle = () => {
>         return "Welcome back " + locals.user.name;
>     };
> 
>     // return a Response or the result of calling `next()`
>     // optionally, transform the response by modifying the result
>     return next();

Reviewer: ```suggestion
    // optionally, modify the properties stored in `locals`
    locals.user.name = "John Wick";
    locals.welcomeTitle = () => {
        return "Welcome back " + locals.user.name;
    };

    // return a Response or the result of calling `next()`
    return next();
```
Just a reminder that whatever we agree on above should also be done here, e.g.
